### PR TITLE
use FileFormat::get_ext as the default file extension filter

### LINKED ISSUE #12378: Consider using the FileFormat::get_ext when creating ListingOptions
### Is your feature request related to a problem or challenge?

When I use the following code to attempt creating external tables using SQL and Rust API respectively, a "Corrupt footer" error occurs. Eventually, I discovered that this was because there were files of other formats in the `partitioned` directory, but DataFusion was reading them as data files. I think this behavior is confusing, because the SQL specifies the Parquet format, and the Rust code also creates a new `ParquetFormat`. 
```sql
CREATE EXTERNAL TABLE hits
STORED AS PARQUET
LOCATION 'partitioned/';
```
```rust
  let mut opts = TableParquetOptions::default();
  opts.set("pushdown_filters", "true").unwrap();
  let format = ParquetFormat::new().with_options(opts);
  let options = ListingOptions::new(Arc::new(format))
      .with_table_partition_cols(vec![("A".to_owned(), DataType::Int32)]);
  ctx.register_listing_table(
      "hits",
      "partitioned/",
      options,
      None,
      None,
  )
  .await
```

### Describe the solution you'd like

Currently, when creating ListingOptions using format, `format.get_ext` is not called. I believe `format.get_ext` should be used as the default file extension filter, which would also better align with the definition of `FileFormat`trait. I you agree with me, I could submit a pr to make that change.

```rust
impl ListingOptions {
  /// Creates an options instance with the given format
  /// Default values:
  /// - no file extension filter
  /// - no input partition to discover
  /// - one target partition
  /// - stat collection
  pub fn new(format: Arc<dyn FileFormat>) -> Self {
      Self {
          file_extension: String::new(),
          format,
          table_partition_cols: vec![],
          collect_stat: true,
          target_partitions: 1,
          file_sort_order: vec![],
      }
  }
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by waruto210**
## Which issue does this PR close?

Closes #12378.

## Rationale for this change

Using `FileFormat::get_ext` as the default file extension filter better meets user expectations.

## What changes are included in this PR?

Use `FileFormat::get_ext` as the default file extension filter

## Are these changes tested?

Yes

## Are there any user-facing changes?

User do not need to specify the file extension if they choose the default value.

### CODE REVIEW

File: datafusion/core/src/datasource/listing/table.rs [hunk 1]
```diff
@@ -1430,15 +1476,18 @@ mod tests {
...
+        file_ext: Option<&str>,
     ) -> Result<()> {
         let ctx = SessionContext::new();
         register_test_store(&ctx, &files.iter().map(|f| (*f, 10)).collect::<Vec<_>>());
 
         let format = AvroFormat {};
 
-        let opt = ListingOptions::new(Arc::new(format))
-            .with_file_extension("")
+        let mut opt = ListingOptions::new(Arc::new(format))
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:1486]: Maybe we could make this a nicer API by adding `with_file_extension_opt` or something so this code could be like:

```rust
let opt = ListingOptions::new(Arc::new(format))
            .with_file_extension_opt(file_ext)
            .with_target_partitions(target_partitions);
```

**Review by alamb — APPROVED** [review]
Thank you @waruto210  -- this makes sense to me and I think is a nice improvement. 

I have a small API improvement suggestion, but I think we can do it as a follow on PR if you prefer.

Thanks again!

**Comment by alamb** [comment]
BTW I am holding off on merging this PR until we cut the 42 release candidate to avoid introducing last minute behavior changes

File: datafusion/core/src/datasource/listing/table.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/datasource/listing/table.rs:1486]: Here is a PR with the proposed API: https://github.com/apache/datafusion/pull/12461

Comment (waruto210) [reply datafusion/core/src/datasource/listing/table.rs]: > Here is a PR with the proposed API: #12461

I have modified the code according to this API.

**Review by alamb — APPROVED** [review]
Thanks @waruto210  -- this is great. 🚀

**Pull request merged**